hal/tests/TESTS/pin_names/pinvalidate.py

```diff
@@ -0,0 +1,307 @@
+            realval = val
+            depth = 0
+            while not re.match("(0x[0-9a-fA-F]+|[1-9][0-9]*|0[1-7][0-7]+|0b[01]+)[uUlL]{0,2}", realval):
+                try:
+                    realval = pinName_dict[realval]
+                    depth += 1
+                except:
+                    break
+                if depth > 10:
+                    break
+
+            if realval in used_pins:
+                message = "already assigned to " + used_pins_friendly[used_pins.index(realval)]
+                invalid_items.append({'key': key, 'val': val, 'message': message})
+                continue
+
+            used_pins.append(realval)
+            used_pins_friendly.append(key + ' = ' + val)
+    return invalid_items
+
+def arduino_existence_check(pinName_dict):
+    required_pins = [
+        'ARDUINO_UNO_A0',
+        'ARDUINO_UNO_A1',
+        'ARDUINO_UNO_A2',
+        'ARDUINO_UNO_A3',
+        'ARDUINO_UNO_A4',
+        'ARDUINO_UNO_A5',
+        'ARDUINO_UNO_D0',
+        'ARDUINO_UNO_D1',
+        'ARDUINO_UNO_D2',
+        'ARDUINO_UNO_D3',
+        'ARDUINO_UNO_D4',
+        'ARDUINO_UNO_D5',
+        'ARDUINO_UNO_D6',
+        'ARDUINO_UNO_D7',
+        'ARDUINO_UNO_D8',
+        'ARDUINO_UNO_D9',
+        'ARDUINO_UNO_D10',
+        'ARDUINO_UNO_D11',
+        'ARDUINO_UNO_D12',
+        'ARDUINO_UNO_D13',
+        'ARDUINO_UNO_D14',
+        'ARDUINO_UNO_D15'
+    ]
+
+    invalid_items = []
+
+    for pin in required_pins:
+        if pin not in pinName_dict:
+            message = pin + " not defined"
+            invalid_items.append({'key': pin, 'val': '', 'message': message})
+    
+    return invalid_items
+
+def arduino_nc_assignment_check(pinName_dict):
+    invalid_items = []
+    for key, val in pinName_dict.items():
+        if re.match("^ARDUINO_UNO_[AD]\d+$", key):
+            if val == 'NC':
+                message = "cannot be NC"
+                invalid_items.append({'key': key, 'val': val, 'message': message})
+    return invalid_items
+
+def legacy_assignment_check(pinName_content):
+    invalid_items = []
+    legacy_assignments = dict(re.findall("^\s*((?:LED|BUTTON)\d*)\s*=\s*([a-zA-Z0-9_]+)", pinName_content, re.MULTILINE))
+    for key, val in legacy_assignments.items():
+        message = "legacy assignment; LEDs and BUTTONs must be #define'd"
+        invalid_items.append({'key': key, 'val': val, 'message': message})
+    return invalid_items
+
+
+def print_summary(report):
+    targets = []
+    for case in report:
+        targets.append(case['platform_name'])
+    targets = list(dict.fromkeys(targets))
+    
+    table = []
+    for target in targets:
+        result = 'PASSED'
+        error_count = 0
+        for case in report:
+            if case['platform_name'] != target:
+                continue
+            if case['result'] == 'FAILED':
+                result = 'FAILED'
+                error_count += 1
+        table.append([target, result, error_count])
```

Comment:
When specifying the path instead of the target, can we shorten the info in the table to make it more readable?
E.g. from 
`.\targets\TARGET_STM\TARGET_STM32L4\TARGET_STM32L4S5xI\TARGET_B_L4S5I_IOT01A\PinNames.h`
To:
`TARGET_B_L4S5I_IOT01A\PinNames.h`

Reply:
Done